groth16/src/json_utils.rs

```diff
@@ -0,0 +1,357 @@
+        let parsed = Self::parse_g2(e);
+        format!(
+            "[[\"{}\", \"{}\"], [\"{}\", \"{}\"]]",
+            parsed.0, parsed.1, parsed.2, parsed.3,
+        )
+    }
+    fn to_g1(x: &str, y: &str) -> Self::G1Affine;
+    fn to_g2(x0: &str, x1: &str, y0: &str, y1: &str) -> Self::G2Affine;
+}
+
+pub fn render_scalar_to_hex<F: PrimeField>(el: &F) -> String {
+    let mut buff = vec![];
+    let repr = el.into_repr();
+    repr.write_be(&mut buff).unwrap();
+
+    format!("0x{}", hex::encode(buff))
+}
+
+pub fn render_hex_to_scalar<F: PrimeField>(value: &str) -> F {
+    let value = BigUint::from_str_radix(&value[2..], 16)
+        .unwrap()
+        .to_str_radix(10);
+    F::from_str(&value).unwrap()
+}
+
+impl Parser for Bn256 {
+    fn parse_g1(e: &Self::G1Affine) -> (String, String) {
+        let (x, y) = e.into_xy_unchecked();
+        (render_scalar_to_hex(&x), render_scalar_to_hex(&y))
+    }
+
+    fn parse_g2(e: &Self::G2Affine) -> (String, String, String, String) {
+        let (x, y) = e.into_xy_unchecked();
+        (
+            render_scalar_to_hex(&x.c0),
+            render_scalar_to_hex(&x.c1),
+            render_scalar_to_hex(&y.c0),
+            render_scalar_to_hex(&y.c1),
+        )
+    }
+
+    fn to_g1(x: &str, y: &str) -> Self::G1Affine {
+        G1Affine::from_xy_unchecked(render_hex_to_scalar(x), render_hex_to_scalar(y))
+    }
+
+    fn to_g2(x0: &str, x1: &str, y0: &str, y1: &str) -> Self::G2Affine {
+        let x = Fq2 {
+            c0: render_hex_to_scalar(x0),
+            c1: render_hex_to_scalar(x1),
+        };
+        let y = Fq2 {
+            c0: render_hex_to_scalar(y0),
+            c1: render_hex_to_scalar(y1),
+        };
+        G2Affine::from_xy_unchecked(x, y)
+    }
+}
+
+impl Parser for Bls12 {
+    fn parse_g1(e: &Self::G1Affine) -> (String, String) {
+        let (x, y) = e.into_xy_unchecked();
+        (render_scalar_to_hex(&x), render_scalar_to_hex(&y))
+    }
+
+    fn parse_g2(e: &Self::G2Affine) -> (String, String, String, String) {
+        let (x, y) = e.into_xy_unchecked();
+        (
+            render_scalar_to_hex(&x.c0),
+            render_scalar_to_hex(&x.c1),
+            render_scalar_to_hex(&y.c0),
+            render_scalar_to_hex(&y.c1),
+        )
+    }
+
+    fn to_g1(x: &str, y: &str) -> Self::G1Affine {
+        G1Affine_bls12381::from_xy_unchecked(render_hex_to_scalar(x), render_hex_to_scalar(y))
+    }
+
+    fn to_g2(x0: &str, x1: &str, y0: &str, y1: &str) -> Self::G2Affine {
+        let x = Fq2_bls12381 {
+            c0: render_hex_to_scalar(x0),
+            c1: render_hex_to_scalar(x1),
+        };
+        let y = Fq2_bls12381 {
+            c0: render_hex_to_scalar(y0),
+            c1: render_hex_to_scalar(y1),
+        };
+        G2Affine_bls12381::from_xy_unchecked(x, y)
+    }
+}
+
+pub fn serialize_vk<P: Parser>(vk: VerifyingKey<P>, curve_type: &str) -> String {
+    format!(
+        "{{
+\"protocol\": \"groth16\",
+\"curve\": \"{}\",
```

Comment:
Since you have defined proof_file and vk file struct, you are serialize the struct to json directly via adding Serialize trait.